pkg/networkserver/grpc.go

```diff
@@ -17,13 +17,131 @@ package networkserver
+	},
+	"factory_preset_frequencies": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		freqs := make([]uint64, 0, len(phy.UplinkChannels))
+		for _, ch := range phy.UplinkChannels {
+			freqs = append(freqs, ch.Frequency)
+		}
+		settings.FactoryPresetFrequencies = freqs
+	},
+	"desired_rx2_frequency": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredRx2Frequency = &ttnpb.FrequencyValue{Value: mac.DeviceDesiredRX2Frequency(nil, phy, fp, nsDefaults)}
+	},
+	"desired_max_duty_cycle": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredMaxDutyCycle = &ttnpb.AggregatedDutyCycleValue{Value: mac.DeviceDesiredMaxDutyCycle(nil, nsDefaults)}
+	},
+	"desired_adr_ack_limit_exponent": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredADRAckLimitExponent = mac.DeviceDesiredADRAckLimitExponent(nil, phy, nsDefaults)
+	},
+	"desired_adr_ack_delay_exponent": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredADRAckDelayExponent = mac.DeviceDesiredADRAckDelayExponent(nil, phy, nsDefaults)
+	},
+	"desired_ping_slot_data_rate_index": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredPingSlotDataRateIndex = mac.DeviceDesiredPingSlotDataRateIndexValue(nil, phy, fp, nsDefaults)
+	},
+	"desired_ping_slot_frequency": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredPingSlotFrequency = &ttnpb.FrequencyValue{Value: mac.DeviceDesiredPingSlotFrequency(nil, phy, fp, nsDefaults)}
+	},
+	"desired_beacon_frequency": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredBeaconFrequency = &ttnpb.FrequencyValue{Value: mac.DeviceDesiredBeaconFrequency(nil, nsDefaults)}
+	},
+	"desired_max_eirp": func(settings *ttnpb.MACSettings, fp *frequencyplans.FrequencyPlan, phy *band.Band, nsDefaults ttnpb.MACSettings) {
+		settings.DesiredMaxEirp = &ttnpb.DeviceEIRPValue{Value: lorawan.Float32ToDeviceEIRP(mac.DeviceDesiredMaxEIRP(nil, phy, fp, nsDefaults))}
+	},
+}
+
+func (ns *NetworkServer) GetDefaultMACSettings(ctx context.Context, req *ttnpb.GetDefaultMACSettingsRequest) (*ttnpb.MACSettings, error) {
+	fp, phy, err := FrequencyPlanAndBand(req.FrequencyPlanID, req.LorawanPhyVersion, ns.FrequencyPlans)
+	if err != nil {
+		return nil, err
+	}
+	settings := &ttnpb.MACSettings{}
+	for _, p := range req.GetFieldMask().GetPaths() {
```

Comment:
This entire thing would be a lot less complicated without all that field masking, so I really think we should just get rid of that.